Create libraries

173210 opened the pull request.

xyzz commented:
What does this do and why?

> On 27 Dec 2016, at 12:00, 173210 <notifications@github.com> wrote:
> 
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/vitasdk/vita-toolchain/pull/99
> 
> Commit Summary
> 
> Create libraries
> File Changes
> 
> M build.sh (6)
> M src/CMakeLists.txt (65)
> A src/build/vita-public.h (9)
> M src/vita-elf-export.c (6)
> M src/vita-export.h (9)
> M src/vita-import.h (29)
> A src/vita-public.h (7)
> M src/yamltree.h (7)
> M src/yamltreeutil.h (19)
> Patch Links:
> 
> https://github.com/vitasdk/vita-toolchain/pull/99.patch
> https://github.com/vitasdk/vita-toolchain/pull/99.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

173210 commented:
This change creates the following libraries: `vita-yaml`, `vita-import` and `vita-export`.
Creating libraries give the following benefits:
* Executables can share binaries and reduce the size in disk and RAM.
* Other softwares such as my [vita-analyze](https://github.com/173210/vita-analyze) can use the infrastructure provided by `vita-import` and `vita-export`.

devnoname120 commented:
Please fix this: https://travis-ci.org/vitasdk/vita-toolchain/builds/186941910#L1489

devnoname120 commented on src/build/vita-public.h:
```diff
@@ -0,0 +1,9 @@
+#ifndef VITA_PUBLIC
```
I'm not fond of this name because it somehow implies that this is vita-related, which it isn't. Maybe `FUNC_PUBLIC` instead?

173210 replied:
The macro is intended to be public and should not conflict with others.

devnoname120 replied:
Not sure what to name it then. It's confusing in some situations like:
```C
VITA_PUBLIC vita_export_t *vita_exports_load(const char *filename, const char *elf, int verbose);
```

173210 replied:
You may name such as `VITA_TOOLCHAIN_PUBLIC`, but I think it's obvious enough if you see the header.

173210 replied:
I mean `VITA_PUBLIC` is enough.

devnoname120 replied:
This seems a bit verbose but I guess this will do.

devnoname120 commented on src/build/vita-public.h:
```diff
@@ -0,0 +1,9 @@
+#ifndef VITA_PUBLIC
+#ifndef VITA_SHARED
```
`FUNC_SHARED`

173210 commented:
2e1faaefbd232b38de8736b0af1863c6be65fd9c: Fixed a silly typo; now it looks fine: https://travis-ci.org/vitasdk/vita-toolchain/builds/187600688.

devnoname120 commented on build.sh:
```diff
@@ -3,18 +3,22 @@ uname_string=`uname | sed 'y/LINUXDARWINFREEOPENPCBSDMSYS/linuxdarwinfreeopenpcb
     JOBS="-j1"
     ;;
   *msys*)
+    BUILD_SHARED_LIBS=ON
```
Why only for msys?

173210 replied:
Because Windows could use shared libraries in the same directory by default.
[Dynamic-Link Library Search Order (Windows) Standard Search Order for Desktop Applications](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682586(v=vs.85).aspx#standard_search_order_for_desktop_applications)

devnoname120 replied:
OK, maybe add a comment so that people know why in the future.

devnoname120 commented on src/vita-public.h:
```diff
@@ -0,0 +1,7 @@
+#else
+#define VITA_PUBLIC
+#endif
+#endif
```
It's not clear that `src/vita-public.h` means that it's for importing whereas `src/build/vita-public.h` is for exporting. I'm not sure why the folder `build` was created by the way.

173210 replied:
Separating directory is required. `build` is the name used in CMake.

devnoname120 commented:
Great, did you try to compile a sample homebrew to make sure that it works?

173210 commented:
Yes, it compiles as expected.

devnoname120 commented:
I don't have much more to comment. I will ask people more involved with the building system than me to review it.

173210 commented:
aa1f387930ec853b267911ebc02d457a8d2b887e
* Renamed `VITA_PUBLIC` and `VITA_SHARED` to `VITA_TOOLCHAIN_PUBLIC` and `VITA_TOOLCHAIN_SHARED` respectively, according to https://github.com/vitasdk/vita-toolchain/pull/99#discussion_r94196723.
* Added comments about Windows according to https://github.com/vitasdk/vita-toolchain/pull/99#discussion_r94196789.

This change affects to the building system, so adding reviewers who have something to do with it is :+1:. See also https://github.com/173210/vita-buildscripts.

devnoname120 commented:
@173210 Nobody else seems to have taken a look. If you feel that it's ready for merging, I will merge it.

173210 commented:
Yes, it's ready.

devnoname120 commented:
It fails to build: https://travis-ci.org/vitasdk/vita-toolchain/builds/192608533

173210 commented:
I have never had the problem. I'm investigating...

173210 commented:
Reproduced in my machine.
